pkg/kv/kvclient/kvcoord/dist_sender_test.go

```diff
@@ -5130,6 +5134,188 @@ func TestSendToReplicasSkipsStaleReplicas(t *testing.T) {
 	}
 }
 
+// Test that DistSender.computeReadAndWriteRUMultipliers() returns the right
+// computed read and write RU multipliers.
+func TestDistSenderComputeRUMultipliers(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	ctx := context.Background()
+	stopper := stop.NewStopper()
+	defer stopper.Stop(ctx)
+
+	clock := hlc.NewClockWithSystemTimeSource(time.Nanosecond /* maxOffset */)
+	rpcContext := rpc.NewInsecureTestingContext(ctx, clock, stopper)
+	rddb := MockRangeDescriptorDB(func(key roachpb.RKey, reverse bool) (
+		[]roachpb.RangeDescriptor, []roachpb.RangeDescriptor, error,
+	) {
+		// computeReadAndWriteRUMultipliers should not be using this at all,
+		// but DistSender insists on having a non-nil one.
+		return nil, nil, errors.New("range desc db unexpectedly used")
+	})
+	st := cluster.MakeTestingClusterSettings()
+
+	// Set regional cost multiplier table.
+	//                     | us-east1 | eu-central1 | asia-southeast1
+	//     -----------------------------------------------------------
+	//        us-east1     |    1     |     1.5     |        2
+	//       eu-central1   |   1.5    |      1      |       3.5
+	//     asia-southeast1 |    2     |     3.5     |        1
+	tenantcostmodel.RegionalCostMultiplierTableSetting.Override(ctx, &st.SV,
+		`{"regions":["us-east1","eu-central1","asia-southeast1"],"matrix":[[1,1.5,2.6],[1,3.5],[1]]}`)
+
+	modelCfg := tenantcostmodel.ConfigFromSettings(&st.SV)
+
+	newRangeDescriptor := func(numReplicas int) *roachpb.RangeDescriptor {
+		desc := &roachpb.RangeDescriptor{
+			InternalReplicas: make([]roachpb.ReplicaDescriptor, numReplicas),
+		}
+		// ReplicaIDs are always NodeIDs + 1 for this test.
+		for i := 1; i <= numReplicas; i++ {
+			desc.InternalReplicas[i-1].NodeID = roachpb.NodeID(i)
+			desc.InternalReplicas[i-1].ReplicaID = roachpb.ReplicaID(i + 1)
+		}
+		return desc
+	}
+
+	makeNodeDescriptor := func(nodeID int, region string) roachpb.NodeDescriptor {
+		return roachpb.NodeDescriptor{
+			NodeID:  roachpb.NodeID(nodeID),
+			Address: util.UnresolvedAddr{},
+			Locality: roachpb.Locality{
+				Tiers: []roachpb.Tier{
+					{Key: "az", Value: fmt.Sprintf("az%d", rand.Intn(10))},
+					{Key: "region", Value: region},
+					{Key: "dc", Value: fmt.Sprintf("dc%d", rand.Intn(10))},
+				},
+			},
+		}
+	}
+
+	for _, tc := range []struct {
+		name          string
+		cfg           *DistSenderConfig
+		desc          *roachpb.RangeDescriptor
+		curReplica    *roachpb.ReplicaDescriptor
+		expectedRead  tenantcostmodel.RUMultiplier
+		expectedWrite tenantcostmodel.RUMultiplier
+	}{
+		{
```

Comment:
It would be nice to include a few tests cases for complex, but expected scenarios. Here are a cases I think we should cover recommendations:

### Simulate a Global Table Local
client: us-east1
currentReplica: us-east1
replicas: {"us-east1": 3, "eu-central1": 1, "asia-southeast1": 1}
expectedWriteMultipler: (3 * 1) + (1 * 1.5) + (1 * 2.6) = 7.1
expectedReadMultipler: 1

### Write to Global Table from non-primary region
client: asia-southeast1
currentReplica: us-east1
replicas: {"us-east1": 3, "eu-central1": 1, "asia-southeast1": 1}
expectedWriteMultipler: (3 * 1.5) + (1 * 1) + (1 * 3.5) = 12.3
expectedReadMultipler: 3.5

### Read from to Global Table in non-primary region
client: eu-central1
currentReplica: eu-central1
replicas: {"us-east1": 3, "eu-central1": 1, "asia-southeast1": 1}
expectedWriteMultipler: (3 * 2.6) + (1 * 3.5) + (1 * 1) = 9
expectedReadMultipler: 1